README.md
<h1 align="center">
    <br>
    <a href="https://easyappointments.org">
        <img src="https://raw.githubusercontent.com/alextselegidis/easyappointments/develop/docs/images/logo.png" alt="7keema" width="150">
    </a>
    <br>
    7keema
    <br>
</h1>

<br>

<h4 align="center">
    A powerful Open Source Appointment Scheduler that can be installed on your server. 
</h4>

<p align="center">
  <img alt="GitHub" src="https://img.shields.io/github/license/alextselegidis/easyappointments?style=for-the-badge">
  <img alt="GitHub release (latest by date)" src="https://img.shields.io/github/v/release/alextselegidis/easyappointments?style=for-the-badge">
  <img alt="GitHub All Releases" src="https://img.shields.io/github/downloads/alextselegidis/easyappointments/total?style=for-the-badge">
</p>

<p align="center">
  <a href="#about">About</a> •
  <a href="#features">Features</a> •
  <a href="#setup">Setup</a> •
  <a href="#installation">Installation</a> •
  <a href="#license">License</a>
</p>

![screenshot](screenshot.png)

## About

**7keema** is a highly customizable web application that allows customers to book appointments with you 
via a sophisticated web interface. Moreover, it provides the ability to sync your data with Google Calendar so you can 
use them with other services. It is an open source project that you can download and install **even for commercial use**. 
7keema will run smoothly with your existing website as it can be installed in a single folder of the 
server and of course share an existing database.

## Features

The application is designed to be flexible enough so that it can handle any enterprise work flow. 

* Customers and appointments management.
* Services and providers organization.
* Working plan and booking rules.
* Google Calendar synchronization.
* Email notifications system.
* Self hosted installation.
* Translated user interface.
* User community support. 

## Setup

To clone and run this application, you'll need [Git](https://git-scm.com), [Node.js](https://nodejs.org/en/download/) (which comes with [npm](http://npmjs.com)) and [Composer](https://getcomposer.org) installed on your computer. From your command line:

```bash
# Clone this repository
$ git clone https://github.com/alextselegidis/easyappointments.git

# Go into the repository
$ cd easyappointments

# Install dependencies
$ npm install && composer install

# Start the file watcher
$ npm start
```

Note: If you're using Linux Bash for Windows, [see this guide](https://www.howtogeek.com/261575/how-to-run-graphical-linux-desktop-applications-from-windows-10s-bash-shell/) or use `node` from the command prompt.

You can build the files by running `npm run build`. This command will bundle everything to a `build` directory.

## Installation

You will need to perform the following steps to install the application on your server:

* Make sure that your server has Apache/Nginx, PHP and MySQL installed.
* Create a new database (or use an existing one).
* Copy the "easyappointments" source folder on your server.
* Make sure that the "storage" directory is writable.
* Rename the "config-sample.php" file to "config.php" and update its contents based on your environment.
* Open the browser on the 7keema URL and follow the installation guide.

That's it! You can now use 7keema at your will.

You will find the latest release at [easyappointments.org](https://easyappointments.org).
If you have problems installing or configuring the application visit the
[official support group](https://groups.google.com/forum/#!forum/easy-appointments).
You can also report problems on the [issues page](https://github.com/alextselegidis/easyappointments/issues)
and help the development progress.

## License 

Code Licensed Under [GPL v3.0](https://www.gnu.org/licenses/gpl-3.0.en.html) | Content Under [CC BY 3.0](https://creativecommons.org/licenses/by/3.0/)

---

Website [alextselegidis.com](https://alextselegidis.com) &nbsp;&middot;&nbsp;
GitHub [alextselegidis](https://github.com/alextselegidis) &nbsp;&middot;&nbsp;
Twitter [@alextselegidis](https://twitter.com/AlexTselegidis)

###### More Projects On Github
###### ⇾ [Plainpad &middot; Self Hosted Note Taking App](https://github.com/alextselegidis/plainpad)
###### ⇾ [Questionful &middot; Web Questionnaires Made Easy](https://github.com/alextselegidis/questionful)
###### ⇾ [Integravy &middot; Service Orchestration At Your Fingertips](https://github.com/alextselegidis/integravy)


CHANGELOG.md
# 7keema - Changelog

This file contains the code changes that were introduced into each release (starting from v1.1.0) so that is easy for 
developers to maintain and readjust their custom modifications on the main project codebase.

## [1.4.2] - 2021-07-27

### Added 

- #1004: Add support for line breaks when displaying the service description in the frontend.
- #1040: Support all-day events while syncing with Google Calendar.

### Fixed

- #961: Timezone/UX issue: Wrong day is selected when timezone differs by -1 day.
- #966: Secretaries are getting notification emails for providers that are not assigned to them.
- #980: Missing Pacific (and potentially other) timezones.
- #982: The Any-Provider option might lead to double bookings, if all the providers have the same number of appointments for the selected date.
- #986: Managed to replicate appointment hash collisions. 
- #989: Fix Critical mistake resulting in wrong date
- #990: The API availabilities controller throws an error when generating availability for services with multiple attendants.
- #991: Available hours generated with the "Any Provider" option in the booking page, may use the information of a provider that is not assigned to the selected service.
- #993: Add support for PHP8 (vendor packages need to be updated).
- #1000: Small fix for the display of the delete button in table view.
- #1011: Working plan exception - details pane shows incorrect details.
- #1023: Backend calendar table events missing or duplicated.
- #1026: The timepicker sliders do not work when using an iOS device.
- #1029: Enhance SMTP functions of PHPMailer.
- #1043: Unavailable events do not block time from services with multiple attendants.
- #1046: Make sure that saving the modifications of a single break does not cancel any pending break edits.
- #1068: Set minimum service duration field value to honor the value of EVENT_MINIMUM_DURATION.
- #1073: Update PHPMailer dependencies.
- #1074: In case of deletion of one appointment, system sends email to admins anyway even if they have email notifications disabled.
- #1092: Javascript RangeError on appointment change causing disabled calendar dates.

## [1.4.1] - 2020-12-14

### Added 

- #952: Add timezone support in the REST API, when managing users.
- #955: Display confirmation modal when disabling a connected Google Calendar Sync.

### Fixed 

- #945: Google Calendar sync throws an error with all day Google Calendar Events.
- #946: Typo in JavaScript code leads to a broken calendar view, when loading unavailability events with note contents.
- #948: Multiple attendant services may lead to double booking.
- #950: Cannot create provider without services via the API, some values (other endpoints) are optional too.
- #953: Current time indicator in fullcalendar is showing time in local timezone and not in the user selected timezone.
- #954: The password must be provided via the API when creating new users.

## [1.4.0] - 2020-12-09

### Added
 
- #203: Appointment location / 12-hour format / sync notes and location in Google Calendar.
- #221: Fixed/Improved sort breaks increasingly by hour within day.
- #247: Add new system-wide setting for removing the "Any Provider" option of the booking page.
- #251: Automatically populate the appointment end datetime in API.
- #301: Automatically reload the backend calendar events.
- #313: How to set the timezone from the user booking the appointment.
- #365: Only allow appointments for a few weeks in advance.
- #431: Add support for working plan exceptions.
- #471: Add new system-wide setting that enable users to choose the first day of the week.
- #496: Add pagination on every backend page in order to make filter requests faster.
- #501: Integrate script for assets minification.
- #502: Config::DEBUG value toggles the use of normal or minified asset files.
- #546: Add appointment edit link in the backend customers page.
- #550: Multi-Lang Front-End selection popup not working on mobile.
- #551: Front-End booking calendar not syncing with business logic working plan.
- #572: Ensure the database structure is compatible to at least MySQL 5.5.
- #576: Appointment cancelled exception not showing properly.
- #610: Token based authentication for the Rest API.
- #648: Add a warning when customers delete their personal information.
- #655: Creating an appointment requires user to enter their phone number enhancement.
- #659: Automatically detect browser language enhancement.
- #663: Language selector not working under legacy iOS (v.10.3.1).
- #680: Generate new password in the generate_random_string function may create duplicate passwords, plus it is not secure enough.
- #739: Enhance the table view mode by replacing the tables with fullcalendar instances.
- #770: Store customer's language and use it with notifications or when the customer manages and existing appointment.
- #889: Notify admins and secretaries on appointment changes.

### Changed 

- #386: Service price should be optional.
- #428: Enable book advance timeout values in days.
- #568: Sort providers alphabetically in the booking page.
- #745: Add appointment notes preview in the event popover.

### Fixed 

- #171: Google calendar sync - wrong timezone for appointments.
- #195: Fix Google calendar sync activation error (JavaScript).
- #298: Provider availability issue when selecting the "Any Provider" option.
- #396: Start and end time do not update correctly during calendar time selection on iPad (and other Safari based devices).
- #447: Captcha error using docker (500 error).

assets/js/app.js
	/*
Author       : Dreamguys
Template Name: Medifab - Bootstrap Admin Template
Version      : 1.0
*/
(function ($) {
	"use strict"; 
	
	// Sidebar

	! function($) {
		"use strict";
		var Sidemenu = function() {
			this.$menuItem = $("#sidebar-menu a");
		};

		Sidemenu.prototype.init = function() {
			var $this = this;
			$this.$menuItem.on('click', function(e) {
			if ($(this).parent().hasClass("submenu")) {
				e.preventDefault();
			}
			if (!$(this).hasClass("subdrop")) {
				$("ul", $(this).parents("ul:first")).slideUp(350);
				$("a", $(this).parents("ul:first")).removeClass("subdrop");
				$(this).next("ul").slideDown(350);
				$(this).addClass("subdrop");
			} else if ($(this).hasClass("subdrop")) {
				$(this).removeClass("subdrop");
				$(this).next("ul").slideUp(350);
			}
		});
			$("#sidebar-menu ul li.submenu a.active").parents("li:last").children("a:first").addClass("active").trigger("click");
		},
		$.Sidemenu = new Sidemenu;

	}(window.jQuery),


	$(document).ready(function($) {
		
		// Sidebar Initiate
		
		$.Sidemenu.init();

		// Sidebar overlay
		
		var $sidebarOverlay = $(".sidebar-overlay");
		$("#mobile_btn, .task-chat").on("click", function(e) {
			var $target = $($(this).attr("href"));
			if ($target.length) {
				$target.toggleClass("opened");
				$sidebarOverlay.toggleClass("opened");
				$("html").toggleClass("menu-opened");
				$sidebarOverlay.attr("data-reff", $(this).attr("href"));
			}
			e.preventDefault();
		});

		$sidebarOverlay.on("click", function(e) {
			var $target = $($(this).attr("data-reff"));
			if ($target.length) {
				$target.removeClass("opened");
				$("html").removeClass("menu-opened");
				$(this).removeClass("opened");
				$(".main-wrapper").removeClass("slide-nav");
			}
			e.preventDefault();
		});
		
		
		// Select 2
		if ($('.select').length > 0) {
			$('.select').select2({
				minimumResultsForSearch: -1,
				width: '100%'
			});
		} 

		// Modal

		if ($('.modal').length > 0) {
			var modalUniqueClass = ".modal";
			$('.modal').on('shown.bs.modal', function(e) {
				var $element = $(this);
				var $uniques = $(modalUniqueClass + ':visible').not($(this));
				if ($uniques.length) {
					$uniques.modal('hide');
					$uniques.one('hidden.bs.modal', function(e) {
						$element.modal('show');
						$("body").addClass("modal-open");
					});
					return false;
				}
			});
		}

		// Floating Label

		if ($('.floating').length > 0) {
			$('.floating').on('focus blur', function(e) {
				$(this).parents('.form-focus').toggleClass('focused', (e.type === 'focus' || this.value.length > 0));
			}).trigger('blur');
		}

		// Right Sidebar Scroll

		if ($('.msg-list-scroll').length > 0) {
			$('.msg-list-scroll').slimscroll({
				height: '100%',
				color: '#878787',
				disableFadeOut: true,
				borderRadius: 0,
				size: '4px',
				alwaysVisible: false,
				touchScrollStep: 100
			});
			var h = $(window).height() - 124;
			$('.msg-list-scroll').height(h);
			$('.msg-sidebar .slimScrollDiv').height(h);

			$(window).resize(function() {
				var h = $(window).height() - 124;
				$('.msg-list-scroll').height(h);
				$('.msg-sidebar .slimScrollDiv').height(h);
			});
		}

		// Left Sidebar Scroll

		if ($('.slimscroll').length > 0) {
			$('.slimscroll').slimScroll({
				height: 'auto',
				width: '100%',
				position: 'right',
				size: "7px",
				color: '#ccc',
				wheelStep: 10,
				touchScrollStep: 100
			});
			var hei = $(window).height() - 60;
			$('.slimscroll').height(hei);
			$('.sidebar .slimScrollDiv').height(hei);

			$(window).resize(function() {
				var hei = $(window).height() - 60;
				$('.slimscroll').height(hei);
				$('.sidebar .slimScrollDiv').height(hei);
			});
		}

		// Page wrapper height

		if ($('.page-wrapper').length > 0) {
			var height = $(window).height();
			$(".page-wrapper").css("min-height", height);
		}

		$(window).resize(function() {
			if ($('.page-wrapper').length > 0) {
				var height = $(window).height();
				$(".page-wrapper").css("min-height", height);
			}
		});

		// Datetimepicker

		if ($('.datetimepicker').length > 0) {
			$('.datetimepicker').datetimepicker({
				format: 'DD/MM/YYYY'
			});
		}

		// Datatable

		if ($('.datatable').length > 0) {
			$('.datatable').DataTable({
				"bFilter": false,
			});
		}

		// Bootstrap Tooltip

		if ($('[data-toggle="tooltip"]').length > 0) {
			$('[data-toggle="tooltip"]').tooltip();
		}

		// Toggle Button

		if ($('.btn-toggle').length > 0) {
			$('.btn-toggle').click(function() {
				$(this).find('.btn').toggleClass('active');
				if ($(this).find('.btn-success').size() > 0) {
					$(this).find('.btn').toggleClass('btn-success');
				}
			});
		}

		// Mobile Menu

		if ($('.main-wrapper').length > 0) {
			var $wrapper = $(".main-wrapper");
			$('#mobile_btn').click(function() {
				$wrapper.toggleClass('slide-nav');
				$('#chat_sidebar').removeClass('opened');
				$(".dropdown.open > .dropdown-toggle").dropdown("toggle");
				return false;
			});
			$('#open_msg_box').click(function() {
				$wrapper.toggleClass('open-msg-box');
				$('.dropdown').removeClass('open');
				return false;
			});
		}

		// Product thumb images

		if ($('.proimage-thumb li a').length > 0) {
			var full_image = $(this).attr("href");
			$(".proimage-thumb li a").click(function() {
				full_image = $(this).attr("href");
				$(".pro-image img").attr("src", full_image);
				return false;
			});
		}

		// Lightgallery

		if ($('#pro_popup').length > 0) {
			$('#pro_popup').lightGallery({
				thumbnail: true,
				selector: 'a'
			});
		}
		
		if ($('#lightgallery').length > 0) {
			$('#lightgallery').lightGallery({
				thumbnail: true,
				selector: 'a'
			});
		}
		
		// Incoming call popup
		
		if ($('#incoming_call').length > 0) {
			$(window).on('load',function(){
				$('#incoming_call').modal('show');
				$("body").addClass("call-modal");
			});
		}

		// Summernote

		if ($('.summernote').length > 0) {
			$('.summernote').summernote({
				height: 200,
				minHeight: null,
				maxHeight: null,
				focus: false
			});
		}

		// Check all email

		if ($('.checkbox-all').length > 0) {
			$('.checkbox-all').click(function() {
				$('.checkmail').click();
			});
		}
		if ($('.checkmail').length > 0) {
			$('.checkmail').each(function() {
				$(this).click(function() {
					if ($(this).closest('tr').hasClass("checked")) {
						$(this).closest('tr').removeClass('checked');
					} else {
						$(this).closest('tr').addClass('checked');
					}
				});
			});
		}

		// Mail important

		if ($('.mail-important').length > 0) {
			$(".mail-important").click(function() {
				$(this).find('i.fa').toggleClass("fa-star");
				$(this).find('i.fa').toggleClass("fa-star-o");
			});
		}

		if ($('.dropdown-toggle').length > 0) {
			$('.dropdown-toggle').click(function() {
				if ($('.main-wrapper').hasClass('open-msg-box')) {
					$('.main-wrapper').removeClass('open-msg-box');
				}
			});
		}
		
		/* Custom Modal */
		
		if ($('.custom-modal').length > 0) {
			$(".custom-modal .modal-content").prepend('<button data-dismiss="modal" class="close" type="button">×</button>');
		}

		// Custom Backdrop for modal popup

		$('a[data-toggle="modal"]').on('click', function() {
			setTimeout(function() {
				if ($(".modal.custom-modal").hasClass('show')) {
					$(".modal-backdrop").addClass('custom-backdrop');

				}
			}, 500);
		});

		// Task function

		var notificationTimeout;

		//Shows updated notification popup 
		var updateNotification = function(task, notificationText, newClass) {
			var notificationPopup = $('.notification-popup ');
			notificationPopup.find('.task').text(task);
			notificationPopup.find('.notification-text').text(notificationText);
			notificationPopup.removeClass('hide success');
			// If a custom class is provided for the popup, add It
			if (newClass)
				notificationPopup.addClass(newClass);
			// If there is already a timeout running for hiding current popup, clear it.
			if (notificationTimeout)
				clearTimeout(notificationTimeout);
			// Init timeout for hiding popup after 3 seconds
			notificationTimeout = setTimeout(function() {
				notificationPopup.addClass('hide');
			}, 3000);
		};

		// Adds a new Task to the todo list 
		var addTask = function() {
			// Get the new task entered by user
			var newTask = $('#new-task').val();
			// If new task is blank show error message
			if (newTask === '') {
				$('#new-task').addClass('error');
				$('.new-task-wrapper .error-message').removeClass('hidden');
			} else {
				var todoListScrollHeight = $('.task-list-body').prop('scrollHeight');
				// Make a new task template
				var newTemplate = $(taskTemplate).clone();
				// update the task label in the new template
				newTemplate.find('.task-label').text(newTask);
				// Add new class to the template
				newTemplate.addClass('new');
				// Remove complete class in the new Template in case it is present
				newTemplate.removeClass('completed');
				//Append the new template to todo list
				$('#task-list').append(newTemplate);
				// Clear the text in textarea
				$('#new-task').val('');
				// As a new task is added, hide the mark all tasks as incomplete button & show the mark all finished button
				$('#mark-all-finished').removeClass('move-up');
				$('#mark-all-incomplete').addClass('move-down');
				// Show notification
				updateNotification(newTask, 'added to list');
				// Smoothly scroll the todo list to the end
				$('.task-list-body').animate({
					scrollTop: todoListScrollHeight
				}, 1000);
			}
		};

		// Closes the panel for entering new tasks & shows the button for opening the panel
		var closeNewTaskPanel = function() {
			$('.add-task-btn').toggleClass('visible');
			$('.new-task-wrapper').toggleClass('visible');
			if ($('#new-task').hasClass('error')) {
				$('#new-task').removeClass('error');
				$('.new-task-wrapper .error-message').addClass('hidden');
			}
		};

		// Initalizes HTML template for a given task 
		//var taskTemplate = $($('#task-template').html());
		var taskTemplate = '<li class="task"><div class="task-container"><span class="task-action-btn task-check"><span class="action-circle large complete-btn" title="Mark Complete"><i class="material-icons">check</i></span></span><span class="task-label" contenteditable="true"></span><span class="task-action-btn task-btn-right"><span class="action-circle large" title="Assign"><i class="material-icons">person_add</i></span> <span class="action-circle large delete-btn" title="Delete Task"><i class="material-icons">delete</i></span></span></div></li>';
		// Shows panel for entering new tasks
		$('.add-task-btn').click(function() {
			var newTaskWrapperOffset = $('.new-task-wrapper').offset().top;
			$(this).toggleClass('visible');
			$('.new-task-wrapper').toggleClass('visible');
			// Focus on the text area for typing in new task
			$('#new-task').focus();
			// Smoothly scroll to the text area to bring the text are in view
			$('body').animate({
				scrollTop: newTaskWrapperOffset
			}, 1000);
		});

		// Deletes task on click of delete button
		$('#task-list').on('click', '.task-action-btn .delete-btn', function() {
			var task = $(this).closest('.task');
			var taskText = task.find('.task-label').text();
			task.remove();
			updateNotification(taskText, ' has been deleted.');
		});

		// Marks a task as complete
		$('#task-list').on('click', '.task-action-btn .complete-btn', function() {
			var task = $(this).closest('.task');
			var taskText = task.find('.task-label').text();
			var newTitle = task.hasClass('completed') ? 'Mark Complete' : 'Mark Incomplete';
			$(this).attr('title', newTitle);
			task.hasClass('completed') ? updateNotification(taskText, 'marked as Incomplete.') : updateNotification(taskText, ' marked as complete.', 'success');
			task.toggleClass('completed');
		});

		// Adds a task on hitting Enter key, hides the panel for entering new task on hitting Esc. key
		$('#new-task').keydown(function(event) {
			// Get the code of the key that is pressed
			var keyCode = event.keyCode;
			var enterKeyCode = 13;
			var escapeKeyCode = 27;
			// If error message is already displayed, hide it.
			if ($('#new-task').hasClass('error')) {
				$('#new-task').removeClass('error');
				$('.new-task-wrapper .error-message').addClass('hidden');
			}
			// If key code is that of Enter Key then call addTask Function
			if (keyCode == enterKeyCode) {
				event.preventDefault();
				addTask();
			}
			// If key code is that of Esc Key then call closeNewTaskPanel Function
			else if (keyCode == escapeKeyCode)
				closeNewTaskPanel();

		});

		// Add new task on click of add task button
		$('#add-task').click(addTask);

		// Close new task panel on click of close panel button
		$('#close-task-panel').click(closeNewTaskPanel);

		// Mark all tasks as complete on click of mark all finished button
		$('#mark-all-finished').click(function() {
			$('#task-list .task').addClass('completed');
			$('#mark-all-incomplete').removeClass('move-down');
			$(this).addClass('move-up');
			updateNotification('All tasks', 'marked as complete.', 'success');
		});

		// Mark all tasks as incomplete on click of mark all incomplete button
		$('#mark-all-incomplete').click(function() {
			$('#task-list .task').removeClass('completed');
			$(this).addClass('move-down');
			$('#mark-all-finished').removeClass('move-up');
			updateNotification('All tasks', 'marked as Incomplete.');
		});

		// Dropfiles

		if ($("#drop-zone").length > 0) {
			var dropZone = document.getElementById('drop-zone');
			var uploadForm = document.getElementById('js-upload-form');
			var startUpload = function(files) {
				console.log(files)
			}

			uploadForm.addEventListener('submit', function(e) {
				var uploadFiles = document.getElementById('js-upload-files').files;
				e.preventDefault()

				startUpload(uploadFiles)
			})

			dropZone.ondrop = function(e) {
				e.preventDefault();
				this.className = 'upload-drop-zone';

				startUpload(e.dataTransfer.files)
			}

			dropZone.ondragover = function() {
				this.className = 'upload-drop-zone drop';
				return false;
			}

			dropZone.ondragleave = function() {
				this.className = 'upload-drop-zone';
				return false;
			}

		}
		
		// Coming Soon

		function getTimeRemaining(endtime) {
			var t = Date.parse(endtime) - Date.parse(new Date());
			var seconds = Math.floor((t / 1000) % 60);
			var minutes = Math.floor((t / 1000 / 60) % 60);
			var hours = Math.floor((t / (1000 * 60 * 60)) % 24);
			var days = Math.floor(t / (1000 * 60 * 60 * 24));
			return {
				'total': t,
				'days': days,
				'hours': hours,
				'minutes': minutes,
				'seconds': seconds
			};
		}

		function initializeClock(id, endtime) {
			var clock = document.getElementById(id);
			var daysSpan = clock.querySelector('.days');
			var hoursSpan = clock.querySelector('.hours');
			var minutesSpan = clock.querySelector('.minutes');
			var secondsSpan = clock.querySelector('.seconds');

			function updateClock() {
				var t = getTimeRemaining(endtime);

				daysSpan.innerHTML = t.days;
				hoursSpan.innerHTML = ('0' + t.hours).slice(-2);
				minutesSpan.innerHTML = ('0' + t.minutes).slice(-2);
				secondsSpan.innerHTML = ('0' + t.seconds).slice(-2);

				if (t.total <= 0) {
					clearInterval(timeinterval);
				}
			}

			updateClock();
			var timeinterval = setInterval(updateClock, 1000);
		}

		var deadline = new Date(Date.parse(new Date()) + 256 * 24 * 60 * 60 * 1000);
		if ($("#countdown").length > 0)
			initializeClock('countdown', deadline);

		// Chart
		
		if ($("#areaChart, #bar-example, #donutChart, #area-chart").length > 0) {
			var colors = [
				'#E94B3B',
				'#39C7AA',
				'#1C7EBB',
				'#F98E33',
				'#ad96da'
			];
			var data = [{
						y: '2014',
						a: 50,
						b: 90
					},
					{
						y: '2015',
						a: 65,
						b: 75
					},
					{
						y: '2016',
						a: 50,
						b: 50
					},
					{
						y: '2017',
						a: 75,
						b: 60
					},
					{
						y: '2018',
						a: 80,
						b: 65
					},
					{
						y: '2019',
						a: 90,
						b: 70
					},
					{
						y: '2020',
						a: 100,
						b: 75
					},
					{
						y: '2021',
						a: 115,
						b: 75
					},
					{
						y: '2022',
						a: 120,
						b: 85
					},
					{
						y: '2023',
						a: 145,
						b: 85
					},
					{
						y: '2024',
						a: 160,
						b: 95
					}
				],
				config = {
					data: data,
					xkey: 'y',
					ykeys: ['a', 'b'],
					labels: ['Total Income', 'Total Outcome'],
					fillOpacity: 0.6,
					hideHover: 'auto',
					behaveLikeLine: true,
					resize: true,
					pointFillColors: ['#ffffff'],
					pointStrokeColors: ['black'],
					gridLineColor: '#eef0f2',
					lineColors: ['gray', '#00bf6f']
				};
			config.element = 'area-chart';
			Morris.Area(config);
			Morris.Line({
				lineColors: colors,
				element: 'areaChart',

				data: [{
						y: '6',
						a: 100,
						b: 40,
						c: 70,
						d: 40
					},
					{
						y: '7',
						a: 120,
						b: 60,
						c: 50,
						d: 50
					},
					{
						y: '8',
						a: 120,
						b: 90,
						c: 80,
						d: 60
					},
					{
						y: '9',
						a: 130,
						b: 120,
						c: 120,
						d: 80
					},

				],
				xkey: 'y',
				ykeys: ['a', 'b', 'c', 'd'],
				resize: true,
				labels: ['Target', 'Starbucks', 'test3', 'test4']
			});
			config.element = 'stacked';
			config.stacked = true;
			Morris.Bar({
				lineColors: colors,
				element: 'bar-example',
				data: [{
						y: '2006',
						a: 100,
						b: 90
					},
					{
						y: '2007',
						a: 75,
						b: 65,
						c: 20,
						d: 55
					},
					{
						y: '2008',
						a: 50,
						b: 40,
						c: 10,
						d: 55
					},
					{
						y: '2009',
						a: 75,
						b: 65,
						c: 25,
						d: 55
					},
					{
						y: '2010',
						a: 50,
						b: 40,
						c: 30,
						d: 55
					},
					{
						y: '2011',
						a: 75,
						b: 65,
						c: 60,
						d: 55
					},
					{
						y: '2012',
						a: 100,
						c: 80,
						d: 55
					}
				],
				xkey: 'y',
				ykeys: ['a', 'b', 'c', 'd'],
				resize: true,
				labels: ['test1', 'test2', 'test3', 'test4']
			});
			Morris.Donut({
				element: 'donutChart',
				data: [{
						value: 40,
						label: 'Tasks'
					},
					{
						value: 15,
						label: 'Clients'
					},
					{
						value: 45,
						label: 'Projects'
					},
					{
						value: 30,
						label: 'Employees'
					},
					{
						value: 15,
						label: 'Messages'
					},
				],
				labelColor: '#333',
				resize: true,
				colors: colors
			});
		}
	});
})(jQuery);